### PR TITLE
Ingamepenaltykick

### LINKED ISSUE #137: Penalty Kicks
In-game penalty kicks need to be reviewed and rules need to be studied.
Are we walking through protected areas when we are not kicking team ?
Are we handling game-controller signals properly ?

### PR DESCRIPTION
**Pull request by MaikRe**
## Introduced Changes
Added behaviors to handle in game penalty kick:
When defending:
Goal keeper has position restrictions
Defenders have an offset
Striker should not try and dribble
No players may enter penalty box
When attacking:
Striker assumes position to take penalty shot
No other players may enter penalty box.

To this end behaviors were added, ball state and game state were adjusted, penalty box obstacle creating for path planning was added.

Fixes #137 

## ToDo / Known 

~~Undefined behavior if Striker is outside penalty box when preparing to execute penalty kick as penalty box obstacles would prevent him from entering the penalty box. Undefined as this would only happen if striker is de-localized in ready phase but re-localizes in set or at beginning of playing phase so that he positions himself wrong in ready phase and cannot correct during playing/executing the penalty kick. Must be observed in real games.~~
Striker ignores penalty box when Hulks are kicking team. Penalty kick location should be changed to in order of complexity and iteration:
1. better distance
2. Isometry for angling shots
3. Kick decision based location

~~defend_*_pose and defend_penalty_*_pose could be merged and position_to_defend decided based on game controller state but removing the lines that set the specific ball position for the penalty case seems to break the behavior~~

## Ideas for Next Iterations (Not This PR)

Add option for goal keeper to jump to block goal.

~~Many improvements could be made as this currently has some duplicate but necessary code. This is essentially the precursor to several probably necessary improvements to the behavior architecture. ~~
Important would be to move all instances of evaluating the sub state to one file (game_state_filter?) so that it is not repeatedly evaluated. Also an improved system of action to role assignments in this type of edge case would be convenient.

## How to Test

Test game needed

The lua files ingamepenalty_* provide various scenarios to test the new behavior, final testing will have to be done in a test game to ensure functionality of some components not implemented in behavior simulator (i.e. verifying correctness of head motion commands).

### CODE REVIEW

**Review by schmidma — CHANGES_REQUESTED** [review]
Thanks for tackling this difficult state machine stuff... I have some questions and ideas for improvements. Feel free to ask questions or start a discussion

File: crates/control/src/behavior/defend.rs [hunk 1]
```diff
@@ -107,6 +137,38 @@ fn defend_left_pose(
...
+            ],
+            field_side: ball.field_side,
+            penalty_shot_direction: Default::default(),
+        })
+        .unwrap_or_default();
+
+    let position_to_defend = point![
+        (-field_dimensions.length + field_dimensions.penalty_area_length) / 2.0,
+        role_positions.defender_y_offset
+    ];
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs:161]: Compared to the regular defending position which is just -field_dimensions.length / 2.0 this is offset by half the penalty area length and essentially works as an offset for the defenders so that they remain outside the penalty box

File: crates/control/src/game_state_filter.rs [hunk 2]
```diff
@@ -235,6 +251,7 @@ impl State {
         cycle_start_time: SystemTime,
         ball_detected_far_from_kick_off_point: bool,
         config: &GameStateFilterConfiguration,
+        ball_detected_far_from_penalty_spot: bool,
     ) -> FilteredGameState {
         let is_in_sub_state = matches!(game_controller_state.sub_state, Some(_));
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs:256]: How so? Ball detected far from kick off point and far from any goal are also both boolean so I'm not sure I understand your comment

File: crates/control/src/path_planner.rs [hunk 3]
```diff
@@ -125,6 +125,41 @@ impl PathPlanner {
...
 
+    pub fn with_penalty_box(
+        &mut self,
+        field_to_robot: Isometry2<f32>,
+        field_dimensions: &FieldDimensions,
+        hulks_attacking: bool,
+    ) {
+        let side_factor: f32 = if hulks_attacking { 1.0 } else { -1.0 };
+        let half_penalty = field_dimensions.penalty_area_width / 2.0;
+        let half_field = field_dimensions.length / 2.0;
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/path_planner.rs]: fixed

File: crates/control/src/game_state_filter.rs [hunk 4]
```diff
@@ -264,8 +281,15 @@ impl State {
                     ball_is_free: !opponent_kick_off && !opponent_sub_state,
                 }
             }
-            State::Playing => FilteredGameState::Playing {
-                ball_is_free: !(is_in_sub_state && opponent_is_kicking_team),
+            State::Playing => match game_controller_state.sub_state {
+                Some(SubState::PenaltyKick) => FilteredGameState::Playing {
+                    ball_is_free: !(is_in_sub_state && opponent_is_kicking_team),
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: Yes correct was duplicated

File: crates/control/src/game_state_filter.rs [hunk 5]
```diff
@@ -264,8 +281,15 @@ impl State {
...
+                },
+                _ => FilteredGameState::Playing {
+                    ball_is_free: !(is_in_sub_state
+                        && opponent_is_kicking_team
+                        && !ball_detected_far_from_penalty_spot),
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: Correct, I moved it to the PenaltyKick match case

File: crates/control/src/game_state_filter.rs [hunk 6]
```diff
@@ -264,8 +281,15 @@ impl State {
...
                 }
             }
-            State::Playing => FilteredGameState::Playing {
-                ball_is_free: !(is_in_sub_state && opponent_is_kicking_team),
+            State::Playing => match game_controller_state.sub_state {
+                Some(SubState::PenaltyKick) => FilteredGameState::Playing {
+                    ball_is_free: !(is_in_sub_state && opponent_is_kicking_team),
+                },
+                _ => FilteredGameState::Playing {
+                    ball_is_free: !(is_in_sub_state
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: > This also applies to the `WhistleInSet` above

I did not touch 'WhistleInSet' in this pull request, so I am not sure if you want me to clean up 'WhistleInSet" or merge 'WhistleInSet' and 'Playing' in order to avoid code duplication

File: crates/control/src/ball_state_composer.rs [hunk 7]
```diff
@@ -2,15 +2,21 @@ use color_eyre::Result;
...
+};
 
 pub struct BallStateComposer {
     last_ball_field_side: Side,
 }
 
 #[context]
-pub struct CreationContext {}
+pub struct CreationContext {
+    pub field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs]: Good catch, was based on old work in 'world_state_composer' but no longer necessary since active vision changes

File: crates/control/src/ball_state_composer.rs [hunk 8]
```diff
@@ -41,7 +49,45 @@ impl BallStateComposer {
             context.team_ball,
             context.robot_to_field,
         ) {
-            (PrimaryState::Ready, ..) => None,
+            (PrimaryState::Ready, _, _, Some(robot_to_field)) => {
+                if let Some(game_controller_state) = context.game_controller_state {
+                    match game_controller_state {
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs:52]: Do you mean merge these into the match statement in line 44 or just merging lines 53 and 54 in order to get rid of 'if let' ?

File: crates/control/src/behavior/stand.rs [hunk 9]
```diff
@@ -1,19 +1,32 @@
...
-                    target: robot_to_field.inverse() * Point2::origin(),
-                },
-            })
+            match world_state.game_controller_state {
+                Some(GameControllerState {
+                    sub_state: Some(SubState::PenaltyKick),
+                    ..
+                }) => Some(MotionCommand::Stand {
+                    head: HeadMotion::LookAt {
+                        target: robot_to_field.inverse() * absolute_last_known_ball_position,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/stand.rs:21]: absolute_last_known_ball_position is set by behavior node to be the ball_state.position which is set by the ball_state_composer in this specific situation, making these implicitly the same. The other option is that I pass the field dimensions as a parameter to this function in order to recalculate this point here. Which option would be better?

File: crates/control/src/behavior/defend.rs [hunk 10]
```diff
@@ -158,6 +283,11 @@ fn defend_goal_pose(
                     kicking_team: Team::Opponent,
                 },
             ..
+        })
+        | Some(GameControllerState {
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs:287]: How can I do this since one uses game_phase and the other uses sub_state? Does this work with some kind of bitwise or | operation?

File: crates/control/src/game_state_filter.rs [hunk 11]
```diff
@@ -77,11 +77,27 @@ impl GameStateFilter {
...
+                ) > context.config.distance_to_consider_ball_moved_in_kick_off
+            })
+            .unwrap_or(false);
+
         let filtered_game_state = self.state.construct_filtered_game_state(
             context.game_controller_state,
             context.cycle_time.start_time,
             ball_detected_far_from_kick_off_point,
             context.config,
+            ball_detected_far_from_penalty_spot,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs:100]: done

File: crates/control/src/path_planner.rs [hunk 12]
```diff
@@ -125,6 +125,41 @@ impl PathPlanner {
         );
     }
 
+    pub fn with_penalty_box(
+        &mut self,
+        field_to_robot: Isometry2<f32>,
+        field_dimensions: &FieldDimensions,
+        hulks_attacking: bool,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/path_planner.rs:132]: Changed, but leads to code duplication in walk_to_pose.rs line 64, how can I combine those cases, i.e how can I pass the kicking_team contained in the game controller state as a parameter to the with_penalty_box function?

File: crates/control/src/behavior/defend.rs [hunk 13]
```diff
@@ -107,6 +137,38 @@ fn defend_left_pose(
...
+        .ball
+        .map(|ball| BallState {
+            position: point![
+                -field_dimensions.length / 2.0 + field_dimensions.penalty_marker_distance,
+                0.0
+            ],
+            field_side: ball.field_side,
+            penalty_shot_direction: Default::default(),
+        })
+        .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs]: This seems to also be an artifact from my original attempts that's now unnecessary due to the ball state composer. I've removed it. The defend_left_pose and defend_penalty_left_pose (and respectively right) now only differ by the defend position. Should I combine these and use a match statement based on penalty kick substate to differ the two?

[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs]: I've reduced code duplication in the defender poses by having the defend position x location be determined by a match statement similar to how goal keeper offset is determined when in penalty shootout phase

File: crates/control/src/ball_state_composer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs]: I've merged most of this, please see conversation about duplicate code in ball_state_composer

File: crates/control/src/ball_state_composer.rs [hunk 14]
```diff
@@ -41,7 +49,45 @@ impl BallStateComposer {
...
+                            ..
+                        } => Some(create_ball_state(
+                            robot_to_field.inverse()
+                                * point![
+                                    context.field_dimensions.length / 2.0
+                                        - context.field_dimensions.penalty_marker_distance,
+                                    0.0
+                                ],
+                            context.robot_to_field,
+                            &mut self.last_ball_field_side,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs]: I've changed it so only plus minus is different for the two but this is currently using another match statement across the entire game_controller_state. Please advise on how to clean up lines 63-66 so that I can just check the kicking team without having to involve the entire game_controller_state again

[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs]: Please see new, very much cleaner, code :)

File: crates/control/src/behavior/defend.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs]: Ive merged these so that only one Some() statement exists, but don't know how to merge it further since one case uses sub_state and the other uses game_phase

File: crates/control/src/path_planner.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/path_planner.rs]: Fixed

File: crates/control/src/game_state_filter.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/game_state_filter.rs:100]: I'm not sure, what you fixed. My point was that `ball_detected_far_from_kick_off_point` and `ball_detected_far_from_penalty_spot` should be next to each other

File: crates/control/src/behavior/walk_to_pose.rs [hunk 15]
```diff
@@ -54,6 +56,20 @@ impl<'cycle> WalkPathPlanner<'cycle> {
...
+            ..
+        }) = game_controller_state
+        {
+            planner.with_penalty_box(
+                robot_to_field.inverse(),
+                self.field_dimensions,
+                kicking_team,
+            )
+        };
+
```
Comment (schmidma) [inline crates/control/src/behavior/walk_to_pose.rs:72]: What do you think about moving this higher level behavior decision to outside this call. Something like `should_avoid_penalty_box`?

File: crates/control/src/game_state_filter.rs [hunk 16]
```diff
@@ -77,11 +77,27 @@ impl GameStateFilter {
...
+            .ball_position
+            .map(|ball| {
+                let absolute_ball_position = *context.robot_to_field * ball.position;
+                distance(
+                    &absolute_ball_position,
+                    &Point2::new(
+                        -context.field_dimensions.length / 2.0
+                            + context.field_dimensions.penalty_marker_distance,
+                        0.0,
+                    ),
```
Comment (schmidma) [inline crates/control/src/game_state_filter.rs:90]: extract this to a variable to increase code readability

**Review by schmidma — COMMENTED** [review]
Another iteration of the review

File: crates/control/src/game_state_filter.rs [hunk 17]
```diff
@@ -265,7 +282,11 @@ impl State {
                 }
             }
             State::Playing => FilteredGameState::Playing {
-                ball_is_free: !(is_in_sub_state && opponent_is_kicking_team),
+                ball_is_free: !(opponent_is_kicking_team
+                    && match game_controller_state.sub_state {
+                        Some(SubState::PenaltyKick) => !ball_detected_far_from_penalty_spot,
+                        _ => is_in_sub_state,
+                    }),
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: removed as discussed

File: crates/control/src/game_state_filter.rs (hunk 16 above)
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: extracted

File: crates/control/src/behavior/walk_to_pose.rs (hunk 15 above)
Comment (MaikRe) [reply crates/control/src/behavior/walk_to_pose.rs]: has been changed as discussed

File: crates/control/src/ball_state_composer.rs [hunk 18]
```diff
@@ -40,15 +46,40 @@ impl BallStateComposer {
...
+                    ..
+                }),
+            ) => Some(create_ball_state(
+                robot_to_field.inverse()
+                    * point![
+                        match kicking_team {
+                            Team::Opponent => -1.0,
+                            _ => 1.0,
+                        } * (context.field_dimensions.length / 2.0
+                            - context.field_dimensions.penalty_marker_distance),
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/ball_state_composer.rs]: chopped

File: crates/control/src/game_state_filter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: changed!

File: crates/control/src/behavior/walk_to_penalty_kick.rs [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+pub fn execute(
+    world_state: &WorldState,
+    walk_and_stand: &WalkAndStand,
+    look_action: &LookAction,
+    path_obstacles_output: &mut AdditionalOutput<Vec<PathObstacle>>,
+    field_dimensions: &FieldDimensions,
+) -> Option<MotionCommand> {
+    let robot_to_field = world_state.robot.robot_to_field?;
+    let kick_off_pose = Isometry2::translation(
+        field_dimensions.length / 2.0 - field_dimensions.penalty_area_length,
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/walk_to_penalty_kick.rs]: This is probably too small of a constant value, might need to be increased or parametrized, but is now relative to penalty spot

File: crates/control/src/game_state_filter.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/game_state_filter.rs]: no longer relevant since penalty spot check was removed as discussed

File: crates/control/src/behavior/defend.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs]: changed to increased defensive radius as discussed

File: crates/control/src/behavior/defend.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MaikRe) [reply crates/control/src/behavior/defend.rs]: rule_ball.or(ball).unwrap_or_else(new_at_center)   :D

File: crates/control/src/ball_state_composer.rs [hunk 20]
```diff
@@ -35,22 +42,56 @@ impl BallStateComposer {
...
     pub fn cycle(&mut self, context: CycleContext) -> Result<MainOutputs> {
+        let rule_ball: Option<BallState> = match (
+            context.primary_state,
+            context.ball_position,
+            context.team_ball,
+            context.robot_to_field,
+            context.game_controller_state,
+        ) {
+            (
+                PrimaryState::Ready | PrimaryState::Set,
```
Comment (schmidma) [inline crates/control/src/ball_state_composer.rs:53]: Only spawn a ball in set, if there is no ball detected

File: crates/control/src/behavior/defend.rs [hunk 21]
```diff
@@ -85,25 +94,31 @@ fn defend_left_pose(
...
-    let defend_pose = block_on_circle(ball.position, position_to_defend, distance_to_target);
+    if let Some(GameControllerState {
+        kicking_team: Team::Opponent,
+        sub_state: Some(SubState::PenaltyKick),
+        ..
+    }) = world_state.game_controller_state
+    {
+        let half_penalty_width = field_dimensions.penalty_area_width / 2.0;
+        let minimum_penalty_defender_radius =
+            (half_penalty_width.powf(2.0) + field_dimensions.penalty_area_length.powf(2.0)).sqrt();
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/behavior/defend.rs:118]: use the `distance` function in nalgebra

File: crates/control/src/behavior/defend.rs [hunk 22]
```diff
@@ -114,25 +129,69 @@ fn defend_right_pose(
...
+        distance_to_target = distance_to_target.max(minimum_penalty_defender_radius);
+    }
+    let defend_pose = block_on_circle(ball.ball_in_field, position_to_defend, distance_to_target);
+    Some(robot_to_field.inverse() * defend_pose)
+}
```
Comment (schmidma) [inline crates/control/src/behavior/defend.rs:159]: code duplication? move it to a function

File: crates/control/src/behavior/node.rs [hunk 23]
```diff
@@ -164,121 +172,137 @@ impl Behavior {
...
+                        ),
+                    }?,
+                ))
             })
             .unwrap_or_else(|| {
                 panic!(
                     "there has to be at least one action available, world_state: {world_state:#?}",
                 )
             });
+        context.active_action.fill_if_subscribed(|| action.clone());
```
Comment (schmidma) [inline crates/control/src/behavior/node.rs:305]: make the `Action` `Copy` and drop the `clone`

File: crates/control/src/behavior/stand.rs [hunk 24]
```diff
@@ -1,19 +1,39 @@
...
+    field_dimensions: &FieldDimensions,
+) -> Option<MotionCommand> {
     match world_state.robot.primary_state {
         PrimaryState::Initial => Some(MotionCommand::Stand {
             head: HeadMotion::ZeroAngles,
             is_energy_saving: true,
         }),
         PrimaryState::Set => {
             let robot_to_field = world_state.robot.robot_to_field?;
+            let target = match world_state.game_controller_state {
```
Comment (schmidma) [inline crates/control/src/behavior/stand.rs:18]: you may be able to drop it?

File: crates/control/src/rule_obstacle_composer.rs [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+        Ok(MainOutputs {
+            rule_obstacles: rule_obstacles.into(),
+        })
+    }
+}
+
+pub fn create_penalty_box(field_dimensions: &FieldDimensions, kicking_team: Team) -> RuleObstacle {
+    let side_factor: f32 = match kicking_team {
+        Team::Hulks => 1.0,
+        _ => -1.0,
```
Comment (schmidma) [inline crates/control/src/rule_obstacle_composer.rs:53]: explicitly match all patterns

File: crates/types/src/action.rs [hunk 26]
```diff
@@ -1,3 +1,7 @@
+use serde::{Deserialize, Serialize};
+use serialize_hierarchy::SerializeHierarchy;
+
+#[derive(Debug, Clone, SerializeHierarchy, Serialize, Deserialize)]
```
Comment (schmidma) [inline crates/types/src/action.rs:4]: + `Copy`

File: crates/types/src/geometry.rs [hunk 27]
```diff
@@ -549,27 +549,31 @@ impl Circle {
...
 #[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializeHierarchy)]
 pub struct Rectangle {
-    pub top_left: Point2<f32>,
-    pub bottom_right: Point2<f32>,
+    pub min: Point2<f32>,
+    pub max: Point2<f32>,
 }
 
 impl Rectangle {
+    pub fn new_with_center(center: Point2<f32>, size: Vector2<f32>) -> Self {
```
Comment (schmidma) [inline crates/types/src/geometry.rs:557]: `..._and_size`?

File: crates/types/src/path_obstacles.rs [hunk 28]
```diff
@@ -52,6 +53,14 @@ pub struct PathObstacle {
     pub populated_connections: HashSet<usize>,
 }
 
+impl PathObstacle {
+    pub fn new_line(start: Point2<f32>, end: Point2<f32>) -> Self {
+        PathObstacleShape::LineSegment(LineSegment(start, end)).into()
+    }
+    pub fn new_circle(center: Point2<f32>, radius: f32) -> Self {
+        PathObstacleShape::Circle(Circle { center, radius }).into()
+    }
```
Comment (schmidma) [inline crates/types/src/path_obstacles.rs:62]: This is also a `From<Circle>`
